### PR TITLE
Hydrogen Bonding

### PR DESCRIPTION
**Pull request by mjohnson541**
This is the first part of the implementation of a scheme for accounting for hydrogen bonding in thermo.  This provides a framework for defining and manipulating hydrogen bonds.  Hydrogen bonds are allowed under the following constraints:  

1) The bond must be between a H and an N or O
2) The H must be covalently bonded to an N or O
3) An atom can only be hydrogen bonded to one other atom
4) Formation of an individual H-bond in the un-H-bonded compound should create no rings with less than five members
5) Only two H-bonds can exist in a given molecule

The last prevents explosion of configurations that thermo would need evaluated for.

### CODE REVIEW

File: rmgpy/molecule/molecule.py [hunk 1]
```diff
@@ -1455,7 +1468,76 @@ def toAdjacencyList(self, label='', removeH=False, removeLonePairs=False, oldSty
...
+        ONatoms = [a for a in self.atoms if a.isOxygen() or a.isNitrogen()]
+        ONinds = [n for n,a in enumerate(self.atoms) if a.isOxygen() or a.isNitrogen()]
+        
+        for i,atm1 in enumerate(self.atoms):
+            if atm1.atomType != H:
+                continue
+            else:
+                atmc = atm1.bonds.keys()[0] #since its a H it should only have one bond
+                if (atmc.isOxygen() or atmc.isNitrogen()) and not any([b.order==0 for b in atmc.bonds.values()]): #this H can be H-bonded
+                    for k,atm2 in enumerate(ONatoms):
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/molecule/molecule.py]: Good catch!

File: rmgpy/molecule/molecule.py [hunk 2]
```diff
@@ -1455,7 +1468,76 @@ def toAdjacencyList(self, label='', removeH=False, removeLonePairs=False, oldSty
...
+        H = atomTypes['H']
+        ONatoms = [a for a in self.atoms if a.isOxygen() or a.isNitrogen()]
+        ONinds = [n for n,a in enumerate(self.atoms) if a.isOxygen() or a.isNitrogen()]
+        
+        for i,atm1 in enumerate(self.atoms):
+            if atm1.atomType != H:
+                continue
+            else:
+                atmc = atm1.bonds.keys()[0] #since its a H it should only have one bond
+                if (atmc.isOxygen() or atmc.isNitrogen()) and not any([b.order==0 for b in atmc.bonds.values()]): #this H can be H-bonded
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/molecule/molecule.py]: No, sorry, the check that the O/N wasn't already bonded to an H was unnecessary here, but the comment is correct.

File: rmgpy/molecule/molecule.py [hunk 3]
```diff
@@ -600,7 +604,14 @@ def isBenzene(self):
...
         """
         return self.isOrder(1.5)
-
+    
+    def isHydrogenBond(self):
+        """
+        Return ``True`` if the bond represents a hydrogen bond or ``False`` if
+        not.
+        """
+        return self.isOrder(0)
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/molecule/molecule.py]: Should be fixed!

**Comment by mjohnson541** [comment]
You can convert them to groups and draw them using pydot, but I don't think drawing them like our nice molecular structures is trivial at all.

**Comment by mjohnson541** [comment]
I think I've fixed the stuff in your comments now.

**Comment by alongd** [comment]
Regarding naming convention, I think we want our new function to have `lowercase_with_underscores` type of names.
https://github.com/ReactionMechanismGenerator/RMG-Py/wiki/Programming-Style-Guide
(I deviate from this rule only where there's an element in the name, like H here, and I keep it capital.)

**Comment by mjohnson541** [comment]
I believe I've made all of the requested changes.

**Comment by mazeau** [comment]
We're using a bond order of 0 to represent van der waals bonds on the cat branch.  Any ideas on how to resolve this?

**Comment by mliu49** [comment]
Since this only added the framework and we aren't using H bonds yet, I think it will be trivial to change the bond order that we use for them. @mjohnson541 should confirm.

I don't know what a good number to change it to would be though. We could potentially use any negative number, though there might be unanticipated consequences (e.g. if there's code somewhere that assumes bond orders to be positive).

**Comment by mjohnson541** [comment]
I can confirm that, however, (not knowing how you guys used vdw bonds) I think it might be worth asking if vdw and H bonds need to be distinct constructs in RMG.

**Comment by mazeau** [comment]
I was thinking I would just change the vdW bond order to 0.5.  I feel they should be kept separate, but I have no good reason why.

**Comment by mjohnson541** [comment]
My thought is that I can't think of any case where defining a vdw bond between atoms A-B is distinct from defining a hydrogen bond between atoms A-B.  To me it seems that the distinction is a function of what atoms A and B are and not the bond itself.

**Comment by alongd** [comment]
I agree with the last comment, though these are different physical phenomena and perhaps in the future we'd like an easy way to differentiate between them (preferably not by checking whether one on both atoms are within a certain list that keeps growing).
As long as we're consistent, we could use any integer number (even large positive ones).

**Comment by rwest** [comment]
On the cat branch, the `vdW` bonds are exclusively between species and surface sites, to represent physisorption. So for now we'll try letting van der Waals and hydrogen bonds co-exist with bond order 0, but it does make me a bit nervous as we may need them to do different things in some scenarios. 

This PR probably isn't the place but it reminds me I want to have a broader discussion about how RMG and RMG-Cat should co-exist. I'm tagging @cfgoldsmith.

**Pull request merged**